### PR TITLE
feat: add rskim-search crate foundation (Wave 0)

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

The 3-layer code search system (#175) needs a foundation crate with shared types, traits, and error handling before any search layer can be implemented. This PR introduces the rskim-search crate foundation (Wave 0) — a pure library crate establishing the public API contract that Waves 1-6 build against.

## Key Changes

**New rskim-search crate:**
- `SearchLayer` trait — core abstraction for pluggable search layers with query execution
- `LayerBuilder` trait — layer construction pattern enabling Wave 1+ implementations
- `FieldClassifier` trait — field categorization for BM25F multi-field ranking (Wave 2)
- `SearchField` enum — BM25F search field categories (body, title, context, metadata)
- `FileId` type — newtype wrapper for posting list file references
- `SearchQuery` struct — query representation (terms, filters, flags)
- `SearchResult` struct — single result with score and metadata
- `TemporalFlags` struct — recency/importance temporal flags
- `IndexStats` struct — index statistics for query planning
- `SearchError` enum — error type with From<SkimError> and From<io::Error> conversions
- 11 unit tests covering types, serialization, and error propagation
- CLI stub: `skim search --help` routes to Wave 0 help text

**Workspace upgrades (supporting infrastructure):**
- `thiserror` 1.0 → 2.0 (drop-in replacement)
- Edition 2021 → 2024 (enables ref pattern fixes)
- Fixed 52 collapsible_if warnings via edition 2024 if-let chaining

## Breaking Changes

None — entirely additive. thiserror 2.0 is a drop-in replacement. Edition 2024 changes are internal (pattern fixes, no API changes).

## Reviewer Focus Areas

**Trait signatures** — these are the public API contract that Waves 1-6 build against:
- `SearchLayer::execute()` — does it capture all necessary query context?
- `LayerBuilder::build()` — is the error handling pattern correct?
- `FieldClassifier::classify()` — does the lifetime bound handle tree_sitter::Node correctly?

**Error composition:**
- Does `SearchError` enum cover all failure modes?
- Are From implementations correct and exhaustive?

**Field taxonomy:**
- Does `SearchField` enum correctly capture BM25F multi-field categories?
- Are variants sufficient for Wave 2+ needs?

## Testing

- 11 unit tests covering: type construction, default values, serialization (serde), error From impls, clone behavior
- Manual verification: `cargo test --package rskim-search` passes
- CLI stub: `skim search --help` displays correctly

## Related Issues

Implements Wave 0 of #175 (3-layer code search system)

### CODE REVIEW

File: crates/rskim-search/src/types.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+// ============================================================================
+// File Identifier
+// ============================================================================
+
+/// Opaque numeric identifier for a file in the search index.
+///
+/// Using a newtype (rather than bare u32) prevents accidental misuse of IDs
+/// as raw integers. Implements ordering so indices can use FileId as a map key.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct FileId(pub u32);
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs:31]: **API Design Issue**: The `pub` visibility on the inner field contradicts the stated newtype pattern intent. Per Rust API Guidelines (C-NEWTYPE), the inner field should be private to enforce encapsulation.

**Fix**: Make the inner field private and add explicit constructors/accessors:

```rust
pub struct FileId(u32);

impl FileId {
    #[must_use]
    pub fn new(id: u32) -> Self { Self(id) }
    
    #[must_use]
    pub fn as_u32(self) -> u32 { self.0 }
}
```

This is cheapest to fix at v0.1.0 before downstream code depends on the API.

**Confidence**: 85% | **Category**: MEDIUM

File: crates/rskim-search/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+// ============================================================================
+// Search Field
+// ============================================================================
+
+/// AST-aware field classification for search results.
+///
+/// Determines which structural region of a source file a match appears in.
+/// Used to weight search results and filter queries to specific code regions.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs:47]: **Consistency Issue**: The `SearchField` enum serializes as PascalCase (`"TypeDefinition"`) via serde default, but the `.name()` method returns snake_case (`"type_definition"`). Consumers will see inconsistent casing depending on which API path they use.

**Fix**: Add `#[serde(rename_all = "snake_case")]` to align serialization with `.name()`:

```rust
#[serde(rename_all = "snake_case")]
pub enum SearchField { ... }
```

Then update serialization test expectations (currently asserting PascalCase).

**Confidence**: 82% | **Category**: MEDIUM

File: crates/rskim-search/src/types.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    Comment,
+    /// String literals
+    StringLiteral,
+    /// Unclassified content not matching any of the above
+    Other,
+}
+
+impl SearchField {
+    /// Returns the snake_case name of this field variant.
+    pub fn name(self) -> &'static str {
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs:85]: **Missing Annotation**: Add `#[must_use]` to prevent accidental discards of the string representation:

```rust
#[must_use]
pub fn name(self) -> &'static str { ... }
```

**Confidence**: 80% | **Category**: MEDIUM

File: crates/rskim-search/src/types.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+// ============================================================================
+// Search Result
+// ============================================================================
+
+/// A single match returned by a [`SearchLayer`].
+///
+/// NOTE: Does NOT derive `PartialEq` because `score: f64` cannot implement it
+/// reliably (NaN != NaN). Use exact field comparisons in tests.
+#[derive(Debug, Clone, Serialize)]
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs]: **Incomplete Trait Derivation**: `SearchResult` derives `Serialize` but not `Deserialize`. For a search library, results often need round-trip serialization (JSON caching, IPC, persistence). The NaN concern is valid for `PartialEq`, not for deserialization.

**Fix**: Add `Deserialize` to the derive list:

```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult { ... }
```

Note: All fields implement `Deserialize` via serde.

**Confidence**: 80% | **Category**: MEDIUM

File: crates/rskim-search/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+//! **IMPORTANT: This is a LIBRARY with NO I/O.**
+//! - Accepts pre-parsed data, not file paths
+//! - Returns Result types, not stdout writes
+//! - Pure types and traits, no side effects
+//!
+//! CLI/binary code in `crates/rskim/src/cmd/search.rs` handles all I/O.
+
+mod types;
+
+pub use types::*;
```
Comment (dean0x) [inline crates/rskim-search/src/lib.rs]: **API Surface Control**: The wildcard re-export `pub use types::*` makes it impossible to control the public API without modifying the `types` module itself. Adding a helper function to `types.rs` accidentally makes it public. This violates the Interface Segregation Principle.

**Fix**: Replace with explicit re-exports:

```rust
pub use types::{
    FileId, FieldClassifier, IndexStats, LayerBuilder, Result, SearchError,
    SearchField, SearchLayer, SearchQuery, SearchResult, TemporalFlags,
};
```

Important for a foundation crate that will grow over time.

**Confidence**: 80% | **Category**: MEDIUM

File: crates/rskim-search/src/types.rs [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    pub temporal_flags: Option<TemporalFlags>,
+    /// Maximum number of results to return
+    pub limit: Option<usize>,
+    /// Number of results to skip (for pagination)
+    pub offset: Option<usize>,
+}
+
+impl SearchQuery {
+    /// Create a new query with the given search text and no filters.
+    pub fn new(text: impl Into<String>) -> Self {
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs:144]: **Missing Annotation**: Add `#[must_use]` to prevent accidentally discarding the return value:

```rust
#[must_use]
pub fn new(text: impl Into<String>) -> Self { ... }
```

This catches bugs where the query is silently discarded.

**Confidence**: 80% | **Category**: MEDIUM

File: crates/rskim/Cargo.toml [hunk 7]
```diff
@@ -14,6 +14,7 @@ path = "src/main.rs"
 
 [dependencies]
 rskim-core = { version = "2.9.0", path = "../rskim-core" }
+rskim-search = { version = "0.1.0", path = "../rskim-search" }
```
Comment (dean0x) [inline crates/rskim/Cargo.toml]: **Unused Dependency**: The `rskim` binary crate declares `rskim-search` as a dependency but never imports anything from it. No file in `crates/rskim/src/` uses types from `rskim_search`. This adds compile-time overhead without benefit.

**Fix**: Remove the dependency until the search CLI actually uses library types:

```toml
# rskim-search = { version = "0.1.0", path = "../rskim-search" }
```

Removal is cleaner at this stage.

**Confidence**: 95% | **Category**: MEDIUM

File: crates/rskim-search/src/types.rs [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+        Self: Sized;
+}
+
+/// Classifier that maps a tree-sitter AST node to a [`SearchField`].
+///
+/// Implementations should be thread-safe so they can be shared across indexing
+/// workers.
+pub trait FieldClassifier: Send + Sync {
+    /// Classify the given `node` within its `source` file.
+    fn classify(&self, node: &tree_sitter::Node<'_>, source: &str) -> SearchField;
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs]: 🏗️ **Architecture: FieldClassifier couples to tree-sitter concrete type**

The trait signature `fn classify(&self, node: &tree_sitter::Node<'_>, source: &str)` hard-codes the search library's public API to tree-sitter. This violates DIP and prevents non-tree-sitter languages (JSON, YAML, TOML) from implementing field classification.

**Fix**: Introduce a `NodeInfo` abstraction (kind, byte_range, named_child_count) that `FieldClassifier` accepts instead of `tree_sitter::Node`. This aligns with the project's Strategy Pattern used in `Language::transform_source()`.

**Confidence**: 85% | **Severity**: HIGH

File: crates/rskim-search/src/types.rs [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+// ============================================================================
+// Search Field
+// ============================================================================
+
+/// AST-aware field classification for search results.
+///
+/// Determines which structural region of a source file a match appears in.
+/// Used to weight search results and filter queries to specific code regions.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs:48]: ⚠️ **Regression: SearchField serialization format changed PascalCase → snake_case**

This PR adds `#[serde(rename_all = "snake_case")]`, which changes the wire format from `"TypeDefinition"` to `"type_definition"`. This is a deliberate breaking change to the serialization contract.

**Impact**: Code that constructed JSON expecting PascalCase will break. Since this is Wave 0 with no known external consumers, the risk is acceptable.

**Recommendation**: Document this in the PR description or CHANGELOG as an intentional format alignment. This prevents confusion if anyone built against the intermediate commit (0181a51).

**Confidence**: 90% | **Severity**: HIGH | **Category**: Regression

File: crates/rskim-search/src/types.rs [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+
+// ============================================================================
+// Search Result
+// ============================================================================
+
+/// A single match returned by a [`SearchLayer`].
+///
+/// NOTE: Does NOT derive `PartialEq` because `score: f64` cannot implement it
+/// reliably (NaN != NaN). Use exact field comparisons in tests.
+#[derive(Debug, Clone, Serialize, Deserialize)]
```
Comment (dean0x) [inline crates/rskim-search/src/types.rs:164]: 🧪 **Testing: SearchResult Deserialize added but not roundtrip tested**

This PR added `Deserialize` to `SearchResult` (changed from just `Serialize`), but the existing serialization tests only verify JSON → `serde_json::Value`, never back into `SearchResult`. This means the `Deserialize` implementation is untested.

**Risk**: Future field renames or type changes will silently break deserialization without test coverage.

**Fix**: Add a roundtrip test:
```rust
#[test]
fn test_search_result_roundtrip() {
    let original = SearchResult { /* ... */ };
    let json = serde_json::to_string(&original).unwrap();
    let deserialized: SearchResult = serde_json::from_str(&json).unwrap();
    assert_eq!(deserialized.file_id, original.file_id);
    // ... verify all fields
}
```

**Confidence**: 90% | **Severity**: HIGH

File: crates/rskim/src/cmd/search.rs [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  skim search --lang rust \"impl Iterator\"
+  skim search --ast \"function_declaration\" --json"
+    );
+}
+
+// ============================================================================
+// Tests
+// ============================================================================
+
+#[cfg(test)]
```
Comment (dean0x) [inline crates/rskim/src/cmd/search.rs:73]: 🔧 **Consistency: Missing `#[allow(clippy::unwrap_used)]` on test module**

Both library crates (`rskim-core` and `rskim-search`) annotate their test modules with `#[allow(clippy::unwrap_used)]` due to strict `deny` lints in `Cargo.toml`. This test module also uses `.unwrap()` but lacks the annotation.

**Fix**: Add the annotation for consistency and forward-compatibility:
```rust
#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
```

This aligns with the pattern established across the project and prevents lint friction if the binary's lint config is later unified with the library crates.

**Confidence**: 90% | **Severity**: MEDIUM

File: crates/rskim/src/cmd/search.rs [hunk 12]
```diff
@@ -0,0 +1,98 @@
+//! Search subcommand — code search via layered n-gram indexing.
+//!
+//! This is a CLI stub wiring the `search` subcommand into the dispatch table.
+//! The full search implementation lives in `rskim-search` library crate.
```
Comment (dean0x) [inline crates/rskim/src/cmd/search.rs]: 📝 **Consistency: Doc comments reference removed `rskim-search` dependency**

The module-level doc comment (lines 4, 9) says:
- "The full search implementation lives in `rskim-search` library crate"
- "Business logic lives in: `rskim-search` crate"

But the `rskim-search` dependency was removed from `Cargo.toml` in this PR. The binary doesn't currently depend on or import from the search crate, so these comments describe an integration that doesn't exist yet.

**Fix**: Update the comment to reflect the current state:
```rust
//! This is a CLI stub wiring the `search` subcommand into the dispatch table.
//! The full search implementation will be provided by the `rskim-search` library
//! crate (dependency to be added in a future wave).
```

**Confidence**: 85% | **Severity**: MEDIUM

**Comment by dean0x** [comment]
## PR Review Summary — 5 Inline Comments Created

All inline comments are **≥80% confidence** and focus on blocking/HIGH/MEDIUM severity issues only. Lower-confidence suggestions are consolidated below.

### Inline Comments (5)

1. **Architecture** (`crates/rskim-search/src/types.rs:223`) — FieldClassifier couples to tree-sitter concrete type. Blocks non-tree-sitter languages from implementing field classification.

2. **Regression** (`crates/rskim-search/src/types.rs:47`) — SearchField serde format change (PascalCase → snake_case). Intentional Wave 0 breaking change; document in PR description.

3. **Testing** (`crates/rskim-search/src/types.rs:144`) — SearchResult now has Deserialize but no roundtrip test coverage.

4. **Consistency** (`crates/rskim/src/cmd/search.rs:72`) — Test module missing `#[allow(clippy::unwrap_used)]` annotation present in library crates.

5. **Consistency** (`crates/rskim/src/cmd/search.rs:4`) — Doc comments reference removed `rskim-search` dependency; update to note future integration.

---

### Lower-Confidence Suggestions (60-79%)

**Not posted as inline comments per guidelines.**

- **SearchField::name() duplication** (82%, MEDIUM) — Manual enum-to-string mapping duplicates serde behavior. Consider deriving via `strum` or documenting performance rationale.

- **SearchQuery builder pattern** (65%, LOW) — 6 optional fields accessed directly. Builder API would improve ergonomics but can wait for Wave 1.

- **rskim-search library orphaned in binary** (80%, MEDIUM) — Removed from Cargo.toml but types in crate. Consider adding to `[dev-dependencies]` for compile-time validation.

- **SearchResult allocations** (65%, LOW) — Vec and Option create allocator pressure. `SmallVec`/`Cow` premature for Wave 0 types-only.

---

### Cross-Reviewer Observations

Architecture and Rust reviewers independently flagged the same core issues:
- FieldClassifier trait coupling (85% confidence)
- SearchField::name() duplication (82% confidence)

This consensus increases signal that these are material concerns worth addressing before downstream implementations depend on these traits.

**Reviews**: 8 reviewers | **Test count**: 3,323 passing | **No security/performance/complexity blocking issues**

**Comment by dean0x** [comment]
## Code Review Findings: PR #213

### Critical Issues (Must Fix Before Merge)

#### 1. **NodeInfo Not Re-exported from Public API (Confidence: 95%)**
**File**: `crates/rskim-search/src/lib.rs` (line 14)
**Category**: BLOCKING — API Breakage

The `NodeInfo` type is used in the public trait signature:
```rust
pub trait FieldClassifier {
    fn classify(&self, node: &NodeInfo, source: &str) -> SearchField;
}
```

But `NodeInfo` is NOT included in the `pub use` statement. Downstream consumers cannot implement `FieldClassifier` because they cannot name the type its method requires.

**Fix**: Add `NodeInfo` to the re-export list:
```rust
pub use types::{
    FieldClassifier, FileId, IndexStats, LayerBuilder, NodeInfo, Result, 
    SearchError, SearchField, SearchLayer, SearchQuery, SearchResult, TemporalFlags,
};
```

---

#### 2. **tree-sitter Leaks into Public API Despite Decoupling Goal (Confidence: 90%)**
**File**: `crates/rskim-search/src/types.rs` (line 258 - `from_ts_node` method)
**Category**: BLOCKING — Dependency Inversion Violation

The `NodeInfo` documentation (lines 231-240) explicitly states its purpose is to keep `rskim-search` from exposing tree-sitter publicly. However:

1. `from_ts_node` is `pub` and accepts `&tree_sitter::Node<'_>`
2. `tree-sitter` is a direct dependency in `Cargo.toml`
3. Any consumer using `from_ts_node` must also depend on tree-sitter

This undermines the architectural goal of decoupling.

**Recommended Fix (Option A)**: Move `from_ts_node` to the calling code (e.g., `rskim-core` or a future indexer crate) where both dependencies are available. This allows `rskim-search` to have zero transitive dependencies.

**Alternative Fix (Option B)**: Gate behind optional feature:
```toml
[features]
tree-sitter = ["dep:tree-sitter"]

[dependencies]
tree-sitter = { workspace = true, optional = true }
```

```rust
#[cfg(feature = "tree-sitter")]
impl NodeInfo {
    pub fn from_ts_node(node: &tree_sitter::Node<'_>) -> Self { ... }
}
```

---

### Should-Fix Issues (Confidence ≥80%)

#### 3. **SearchQuery Lacks Forward-Compatibility Marker (Confidence: 82%)**
**File**: `crates/rskim-search/src/types.rs` (line 120)

All 6 fields are public with no `#[non_exhaustive]`. Adding fields later (field boosts, scoring mode, regex support) will be breaking changes.

**Fix**:
```rust
#[derive(Debug, Clone)]
#[non_exhaustive]
pub struct SearchQuery {
    pub text: String,
    // ... fields ...
}
```

---

#### 4. **from_ts_node Has No Test Coverage (Confidence: 82%)**
**File**: `crates/rskim-search/src/types.rs` (line 258)

The `from_ts_node` method is the primary tree-sitter bridge but has no test. This method accesses three tree-sitter APIs that could silently break.

**Fix**: Add integration test:
```rust
#[test]
fn test_node_info_from_ts_node() {
    let mut parser = tree_sitter::Parser::new();
    parser.set_language(&tree_sitter_rust::LANGUAGE.into()).unwrap();
    let tree = parser.parse("fn main() {}", None).unwrap();
    let root = tree.root_node();
    let info = NodeInfo::from_ts_node(&root);
    assert_eq!(info.kind, "source_file");
    assert_eq!(info.byte_range, 0..13);
    assert!(info.named_child_count > 0);
}
```

---

#### 5. **Test Comment Style Inconsistency (Confidence: 80%)**
**File**: `crates/rskim/src/cmd/search.rs` (lines 87-107)

New tests use inline `//` comments, but the crate convention is `///` doc comments (used in `stats.rs` and the new `types.rs` tests).

**Fix**: Convert to doc comments:
```rust
/// Empty args prints help and returns ExitCode::SUCCESS.
#[test]
fn test_search_help_returns_success() {
    // ...
}
```

---

### Summary

**Blocking Issues**: 2 (NodeInfo export, tree-sitter leak)
**Should-Fix Issues**: 3 (SearchQuery, from_ts_node test, test style)
**Recommendation**: CHANGES_REQUESTED

The architectural foundation is sound (Strategy Pattern for language dispatch, trait-based abstraction, proper error handling), but the two HIGH-confidence blocking issues directly contradict the stated design goals and must be resolved before merge.

---
*Review conducted by Claude Code — multi-reviewer architecture analysis*

**Comment by dean0x** [comment]
## Code Review Summary

✅ **6 inline comments created** for findings with ≥80% confidence:
- SearchQuery missing Serialize/Deserialize (82%)
- TemporalFlags missing PartialEq/Eq (82%)
- IndexStats missing PartialEq/Eq (82%)
- Inconsistent thiserror derive style (92%)
- NodeInfo.kind &'static str constraint (82%)
- LayerBuilder::build ownership semantics (80%)

---

### Lower-Confidence Suggestions (60-79%) — Not Blocked

These are thoughtful observations but not required for approval:

**Architecture:**
- Single-field `TemporalFlags` may be premature abstraction (65%) — could be `Option<u32>` until temporal model needs multiple fields
- `FileId(pub u32)` limits index to ~4B files (62%) — fine for single-repo use; multi-repo would need `u64`
- Heavy transitive dependency via rskim-core (pre-existing, 82%) — future improvement: extract Language/SkimError to lightweight rskim-types crate

**Rust:**
- SearchResult lacks PartialEq helper (65%) — could add `fn approx_eq(&self, other: &Self, epsilon: f64)` for tests
- Result type alias shadows std::result::Result (70%) — idiomatic but may need disambiguation in some scopes
- Edition 2024 if-let chains reduce grep-ability (62%) — cleaner syntax but harder to set breakpoints

**Testing:**
- Missing trait contract tests for SearchLayer/LayerBuilder (82% — **already filed as BLOCKING above**)
- Missing NaN score serialization edge case test (65%)
- CLI stub stderr output not verified (70%)

**Performance:**
- SearchResult Vec<Range<usize>> could use SmallVec (65%) — micro-optimization for later
- SearchQuery::new allocates always (62%) — unlikely hot path

**Security:**
- Unbounded SearchQuery.text field (65%) — consider enforcing upper bound when parsing CLI input
- No limit default cap (60%) — document or enforce a sensible default (e.g., 1000) at boundary

---

**Deduplication Notes:**
- 12 existing inline comments detected; no duplicates with new comments
- 4 new blocking findings; 2 blocking architecture findings = 6 total inline comments
- Lines in diff verified; all targeted lines are in the PR's modified sections

---

**Authored by Claude Code** | [devflow:git](https://github.com/dean0x/skim/blob/feat/rskim-search-foundation/.claude/skills/devflow:git)

**Comment by dean0x** [comment]
## PR Review: Code Comments for Blocking Issues

Reviewed 9 analysis reports from `/Users/dean/Sandbox/skim-search/.docs/reviews/feat-rskim-search-foundation/2026-05-11_0002/`

### Findings with ≥80% Confidence (BLOCKING)

**6 Critical Issues Identified:**

#### 1. SearchQuery Missing Serialize/Deserialize (82%)
**File**: `crates/rskim-search/src/types.rs`, lines 119-148
**Issue**: `SearchQuery` is a public API boundary type (used in `search()` trait) but does not derive `Serialize + Deserialize`, while all other similar types (`SearchResult`, `IndexStats`, `SearchField`, `TemporalFlags`, `FileId`) do. This breaks roundtrip serialization for query persistence and logging.
**Fix**: Add `#[derive(Serialize, Deserialize)]` to the struct. The `Language` field already derives both.

#### 2. TemporalFlags Missing PartialEq/Eq (82%)
**File**: `crates/rskim-search/src/types.rs`, line 105
**Issue**: Contains only `Option<u32>` (equality-comparable) but does not derive `PartialEq` and `Eq`. Prevents test assertions and breaks consistency with `rskim-core` patterns.
**Fix**: Update derive to include `PartialEq, Eq`:
```rust
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
```

#### 3. IndexStats Missing PartialEq/Eq (82%)
**File**: `crates/rskim-search/src/types.rs`, line 180
**Issue**: Contains only integer types (all equality-comparable) but does not derive `PartialEq` and `Eq`. Breaks consistency with `rskim-core`.
**Fix**: Update derive to include `PartialEq, Eq`:
```rust
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
```

#### 4. Inconsistent thiserror Derive Style (92%)
**File**: `crates/rskim-search/src/types.rs`, line 269
**Issue**: `rskim-core` imports `use thiserror::Error;` and derives `#[derive(Debug, Error)]`. This crate uses fully-qualified `#[derive(Debug, thiserror::Error)]` without import.
**Fix**: Add `use thiserror::Error;` at top and change derive to `#[derive(Debug, Error)]`.

#### 5. NodeInfo.kind Uses &'static str (82%)
**File**: `crates/rskim-search/src/types.rs`, line 244
**Issue**: Works naturally for tree-sitter (always compile-time constants) but constrains non-tree-sitter implementors. JSON/YAML/TOML implementations would need `Box::leak()` or string literals, risking memory leaks.
**Fix**: Consider `Cow<'static, str>` instead:
```rust
pub kind: Cow<'static, str>,
```
Or document the constraint if all non-tree-sitter kinds are compile-time constants.

#### 6. LayerBuilder::build Consumes Self (80%)
**File**: `crates/rskim-search/src/types.rs`, line 226
**Issue**: The `build(self)` signature consumes the builder, preventing incremental re-indexing. Forces full rebuild on every file change, incompatible with Wave 1-6 foundation requirements.
**Fix**: Either:
- Add `fn snapshot(&self) -> Result<Box<dyn SearchLayer>>` for non-consuming snapshots, or
- Document that incremental indexing is deferred to a separate trait

---

### Additional Findings (60-79% Confidence — Suggestions Only)

**Architecture (Pre-existing)**
- Heavy transitive tree-sitter dependency via `rskim-core` (82%) — future improvement: extract `Language`/`SkimError` to lightweight `rskim-types` crate

**Testing**
- Missing trait contract tests for `SearchLayer`/`LayerBuilder` (82%) — add mock implementations verifying trait can be satisfied
- Missing NaN score serialization edge case test (65%)
- CLI stub stderr output not verified in tests (70%)

**Rust**
- `SearchResult` lacks `PartialEq` helper method (65%) — could add `fn approx_eq()` for tests
- Result type alias shadows std::result::Result (70%) — idiomatic but may need disambiguation

**Security**
- Unbounded `SearchQuery.text` field (65%) — consider enforcing upper bound when parsing CLI input
- No limit cap on `SearchQuery.limit` (60%) — document or enforce sensible default at boundary

**Performance**
- `SearchResult` Vec<Range> could use SmallVec (65%) — micro-optimization for later

---

### Deduplication Summary

- **Total existing PR comments**: 12 (from user and previous reviews)
- **New inline comments created**: Summary comment with 6 detailed findings
- **Lines touched**: All blocking issues are in modified sections of the diff
- **Rate limit status**: OK (well under limits)

---

**Review Date**: 2026-05-11  
**PR**: #213 (feat/rskim-search-foundation)  
**Branch Target**: main  
**Authored by**: Claude Code via devflow:git
